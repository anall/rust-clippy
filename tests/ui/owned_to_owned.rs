#![warn(clippy::owned_to_owned)]
#![allow(clippy::redundant_clone)]
use std::borrow::Borrow;
use std::path::PathBuf;

fn return_owned_from_ref(result : &[u32]) -> Vec<u32> {
    result.to_owned()
}

#[derive(Copy,Clone)]
struct Kitten {}
impl Kitten {}
impl Borrow<BorrowedKitten> for Kitten {
    fn borrow(&self) -> &BorrowedKitten {
        static VALUE: BorrowedKitten = BorrowedKitten {};
        &VALUE
    }
}

struct BorrowedKitten {}
impl ToOwned for BorrowedKitten {
    type Owned = Kitten;
    fn to_owned(&self) -> Kitten {
        Kitten {}
    }
}

fn main() {
    let result = vec![5];
    let _ = return_owned_from_ref(&result);
    let _ = result.to_owned();
    let _ = result.to_vec();

    let str = "hello world".to_string();
    let _ = str.to_owned();

    let kitten = Kitten{};
    let _ = kitten.to_owned();

    let borrowed = BorrowedKitten{};
    let _ = borrowed.to_owned();

    let pathbuf = PathBuf::new();
    let _ = pathbuf.to_owned();
    let _ = pathbuf.to_path_buf();
}
