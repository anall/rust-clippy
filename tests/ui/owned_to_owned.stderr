error: using `to_owned` on an already-owned type
  --> $DIR/owned_to_owned.rs:30:20
   |
LL |     let _ = result.to_owned();
   |                    ^^^^^^^^ help: replace this with: `clone`
   |
   = note: `-D clippy::owned-to-owned` implied by `-D warnings`

error: using `to_vec` on an already-owned type
  --> $DIR/owned_to_owned.rs:31:20
   |
LL |     let _ = result.to_vec();
   |                    ^^^^^^ help: replace this with: `clone`

error: using `to_owned` on an already-owned type
  --> $DIR/owned_to_owned.rs:34:17
   |
LL |     let _ = str.to_owned();
   |                 ^^^^^^^^ help: replace this with: `clone`

error: using `clone` on type `Kitten` which implements the `Copy` trait
  --> $DIR/owned_to_owned.rs:37:13
   |
LL |     let _ = kitten.clone();
   |             ^^^^^^^^^^^^^^ help: try removing the `clone` call: `kitten`
   |
   = note: `-D clippy::clone-on-copy` implied by `-D warnings`

error: aborting due to 4 previous errors

